### PR TITLE
Sieve: Accept :zone +/-hhmm OR :zone iana-tzid

### PR DESCRIPTION
**Pull request by ksmurchison**
Extend the [current]date :zone argument (RFC 5260) to also accept an IANA TZID, from which we will extract the time zone offset

### CODE REVIEW

**Review by elliefm — COMMENTED** [review]
The changes look okay so far, but there's no changes to tests, which means there's can't be any tests for the new iana-tzid behaviour.

Ahh, and there is a failing cunit test:

```
Suite: sieve
  [...]
  Test: date_zone ...FAILED
    1. ./cunit/sieve.testc:1478  - CU_ASSERT_EQUAL(ctx.stats.redirects=0,1=1)
    2. ./cunit/sieve.testc:1479  - CU_ASSERT_EQUAL(ctx.stats.keeps=1,0=0)
    3. ./cunit/sieve.testc:1480  - CU_ASSERT_STRING_EQUAL(ctx.redirected_to="(null)","me@blah.com"="me@blah.com")
    4. ./cunit/sieve.testc:1485  - CU_ASSERT_EQUAL(ctx.stats.redirects=0,1=1)
    5. ./cunit/sieve.testc:1486  - CU_ASSERT_EQUAL(ctx.stats.keeps=2,1=1)
    6. ./cunit/sieve.testc:1487  - CU_ASSERT_STRING_EQUAL(ctx.redirected_to="(null)","me@blah.com"="me@blah.com")
  [...]
```

I haven't investigated it, but off the top of my head I wonder if it's because of the change from minutes to seconds in the internal representation?  Since the `:zone [+-]hhmm` syntax hasn't changed otherwise, so an existing test using that syntax shouldn't have started failing.

On the other hand, Cassandane runs clean, which is interesting -- I guess it doesn't exercise whatever edge case cunit is tripping over.

**Comment by ksmurchison** [comment]
@elliefm Fixed the bug causing the cunit test failures and added Cass tests

**Review by elliefm — APPROVED** [review]
Looks good, and tests all pass now.

One nit where the tz offset parser could be hardened up, but I think hardening it would be a trivial change and doesn't need re-review afterwards.

File: sieve/bc_eval.c [hunk 1]
```diff
@@ -385,6 +389,21 @@ static int do_comparisons(strarray_t *needles, const char *hay,
...
+{
+    int hh, mm;
+    char sign;
+
+    /* Parse +/-hhmm and convert to seconds */
+    if ((3 == sscanf(s, "%c%02d%02d", &sign, &hh, &mm)) &&
+        (sign == '+' || sign == '-') && (mm < 60)) {
+        *offset = (sign == '-' ? -1 : 1) * 60 * (hh * 60 + mm);
+        return 1;
+    }
```
[the remark this answers is not in the document]
Comment (ksmurchison) [reply sieve/bc_eval.c]: Fixed
